hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala

```diff
@@ -142,11 +142,24 @@ case class HoodieResolveReferences(sparkSession: SparkSession) extends Rule[Logi
         val resolvedCondition = condition.map(resolveExpressionFrom(resolvedSource)(_))
         val resolvedAssignments = if (isInsertOrUpdateStar(assignments)) {
           // assignments is empty means insert * or update set *
-          // we fill assign all the source fields to the target fields
-          target.output
-            .filter(attr => !HoodieSqlUtils.isMetaField(attr.name))
-            .zip(resolvedSource.output.filter(attr => !HoodieSqlUtils.isMetaField(attr.name)))
-            .map { case (targetAttr, sourceAttr) => Assignment(targetAttr, sourceAttr) }
+          val resolvedSourceOutputWithoutMetaFields = resolvedSource.output.filter(attr => !HoodieSqlUtils.isMetaField(attr.name))
+          val targetOutputWithoutMetaFields = target.output.filter(attr => !HoodieSqlUtils.isMetaField(attr.name))
+          val resolvedSourceColumnNamesWithoutMetaFields = resolvedSourceOutputWithoutMetaFields.map(attr => attr.name)
+
+          if(targetOutputWithoutMetaFields.filter(attr => resolvedSourceColumnNamesWithoutMetaFields.contains(attr.name)).length
+            == targetOutputWithoutMetaFields.length) {
+            //If sourceTable's columns contains all targetTable's columns,
+            //We fill assign all the source fields to the target fields by column name matching.
+            targetOutputWithoutMetaFields.map(targetAttr => {
+              val sourceAttr = resolvedSourceOutputWithoutMetaFields.filter(attr => attr.name.equals(targetAttr.name)).head
```

Comment:
Can we fetch the sourceAttr by Map?

Reply:
Hello, is that right？

`val sourceColNameAttrMap = resolvedSourceOutputWithoutMetaFields.map(attr => (attr.name,attr)).toMap
val sourceAttr = sourceColNameAttrMap.get(targetAttr.name).get
`